File: src/rb/PhantasyIslandRubyRemoteControl/ph0apy.rb
require_relative 'airplane_manager'

# 此类是到FH0A库的适配器，是对AirplaneManager的wrapper
class FH0A
  def initialize
    @airs = get_airplane_manager
    @airs.flush
    @airs.start
    @airs.flush
  end

  # sleep 单位 秒
  def sleep(time)
    Kernel.sleep(time)
  end

  def destroy
  end

  # 添加（注册）无人机
  def add_uav(port)
    @airs.get_airplane(port)
  end

  def p(port)
    @airs.get_airplane(port)
  end

  # 降落
  def land(port)
    p(port).land
  end

  # 起飞到指定高度 单位cm
  def takeoff(port, high)
    p(port).takeoff(high)
  end

  # 上升指定距离 单位cm
  def up(port, distance)
    p(port).up(distance)
  end

  # 下降指定距离 单位cm
  def down(port, distance)
    p(port).down(distance)
  end

  # 前进指定距离 单位cm
  def forward(port, distance)
    p(port).forward(distance)
  end

  # 后退指定距离 单位cm
  def back(port, distance)
    p(port).back(distance)
  end

  # 左移指定距离 单位cm
  def left(port, distance)
    p(port).left(distance)
  end

  # 右移指定距离 单位cm
  def right(port, distance)
    p(port).right(distance)
  end

  # 移动到指定坐标处
  def goto(port, x, y, h)
    p(port).goto(x, y, h)
  end

  # flip函数用于控制无人机翻滚
  # @param direction 翻滚方向（f前 b后 l左 r右）
  def flip(port, direction)
    case direction
    when 'f'
      p(port).flip_forward
    when 'b'
      p(port).flip_back
    when 'r'
      p(port).flip_right
    when 'l'
      p(port).flip_left
    end
  end

  # 顺时旋转指定角度
  def rotate(port, degree)
    p(port).rotate(degree)
  end

  # 顺时针旋转指定角度
  def cw(port, degree)
    p(port).cw(degree)
  end

  # 逆时针旋转指定角度
  def ccw(port, degree)
    p(port).ccw(degree)
  end

  # 设置飞行速度
  def speed(port, speed)
    p(port).speed(speed)
  end

  # 移动到指定高度处
  def high(port, high)
    p(port).high(high)
  end

  # 设置无人机led色彩
  def led(port, r, g, b)
    p(port).led(r, g, b)
  end

  # 设置无人机led呼吸灯色彩
  def bln(port, r, g, b)
    p(port).bln(r, g, b)
  end

  # 设置无人机led彩虹色彩
  def rainbow(port, r, g, b)
    p(port).rainbow(r, g, b)
  end

  # 设置无人机飞行模式
  # @param mode 1常规2巡线3跟随4单机编队 通常情况下使用模式4
  def mode(port, mode)
    p(port).airplane_mode(mode)
  end

  def color_detect(port, l_l, l_h, a_l, a_h, b_l, b_h)
  end

  def vision_mode(port, mode)
  end

  # 停桨
  def stop(port)
    p(port).stop
  end

  # 悬停
  def hover(port)
    p(port).hover
  end
end

File: src/go/PhantasyIslandGoRemoteControl/ph0apy.go
package PhantasyIslandGoRemoteControl

import (
	"time"
)

// FH0A 此类是到FH0A库的适配器，是对AirplaneManager的wrapper
type FH0A struct {
	Airs *AirplaneManager
}

func NewFH0A() *FH0A {
	f := &FH0A{
		Airs: GetAirplaneManager(),
	}
	f.Airs.Flush()
	f.Airs.Start()
	f.Airs.Flush()
	return f
}

func (f *FH0A) Sleep(seconds float64) {
	time.Sleep(time.Duration(seconds * float64(time.Second)))
}

func (f *FH0A) Destroy() {
	// 实现销毁逻辑
}

func (f *FH0A) AddUav(port string) {
	f.Airs.GetAirplane(port)
}

func (f *FH0A) P(port string) *AirplaneController {
	return f.Airs.GetAirplane(port)
}

func (f *FH0A) Land(port string) {
	f.P(port).Land()
}

func (f *FH0A) Takeoff(port string, high int) {
	f.P(port).Takeoff(high)
}

func (f *FH0A) Up(port string, distance int) {
	f.P(port).Up(distance)
}

func (f *FH0A) Down(port string, distance int) {
	f.P(port).Down(distance)
}

func (f *FH0A) Forward(port string, distance int) {
	f.P(port).Forward(distance)
}

func (f *FH0A) Back(port string, distance int) {
	f.P(port).Back(distance)
}

func (f *FH0A) Left(port string, distance int) {
	f.P(port).Left(distance)
}

func (f *FH0A) Right(port string, distance int) {
	f.P(port).Right(distance)
}

func (f *FH0A) Goto(port string, x int, y int, h int) {
	f.P(port).Goto(x, y, h)
}

func (f *FH0A) Flip(port string, direction string) {
	p := f.P(port)
	switch direction {
	case "f":
		p.FlipForward()
	case "b":
		p.FlipBack()
	case "r":
		p.FlipRight()
	case "l":
		p.FlipLeft()
	}
}

func (f *FH0A) Rotate(port string, degree int) {
	f.P(port).Rotate(degree)
}

func (f *FH0A) Cw(port string, degree int) {
	f.P(port).Cw(degree)
}

func (f *FH0A) Ccw(port string, degree int) {
	f.P(port).Ccw(degree)
}

func (f *FH0A) Speed(port string, speed int) {
	f.P(port).Speed(speed)
}

func (f *FH0A) High(port string, high int) {
	f.P(port).High(high)
}

func (f *FH0A) Led(port string, r, g, b int) {
	f.P(port).Led(r, g, b)
}

func (f *FH0A) Bln(port string, r, g, b int) {
	f.P(port).Bln(r, g, b)
}

func (f *FH0A) Rainbow(port string, r, g, b int) {
	f.P(port).Rainbow(r, g, b)
}

func (f *FH0A) Mode(port string, mode int) {
	f.P(port).AirplaneMode(mode)
}

func (f *FH0A) ColorDetect(port string, LL, LH, AL, AH, BL, BH int) {
	// TODO
}

func (f *FH0A) VisionMode(port string, mode int) {
	// TODO
}

func (f *FH0A) Stop(port string) {
	f.P(port).Stop()
}

func (f *FH0A) Hover(port string) {
	f.P(port).Hover()
}
